HW5 HTML-интерфейс

r-i-besedin opened the pull request:
Перенес все что связано с web в отдельный пакет, добавил таймауты, добавил пользовательский интерфейс, функцию сортировки результата перенес в пакет с индексом, файл с тестами перенес в пакет, который тестируется. Задавать параметры запуска теперь можно через переменные среды. Работу с шаблонами сделал, пока что, через стандартную библиотеку 🙈

r-i-besedin committed:
Добавлен пользовательский HTML интерфейс

Перенес все что связано с web в отдельный пакет, добавил пользовательский интферфейс, функцию сортировки результата перенес в пакет с индексом, файл с тестами перенес в пакет, который тестируется. Задавать параметры запуска теперь можно через переменные среды

tariel-x commented:
Не хватает логирования.
Можно посмотреть примерно здесь: https://github.com/polis-mail-ru-golang-1/examples/blob/master/lec6/blog/main.go#L21

tariel-x commented on main.go:
```diff
@@ -2,37 +2,21 @@ package main
-	searchingfolder := os.Args[1] 	//"./search"
-	interfaceAddr := os.Args[2]		//"127.0.0.1:8080"
+	searchingfolder := os.Getenv("SDIR") //"./search"
+	interfaceAddr := os.Getenv("ADDR")	//"127.0.0.1:8080"
```
Здорово было бы вынести это в отдельный пакет или хотя бы отдельную структуру, содержащую конфиг. И определить значение по-умолчанию для ADDR. Пример: https://github.com/polis-mail-ru-golang-1/examples/blob/master/lec6/blog/config/config.go

tariel-x commented on web/web.go:
```diff
@@ -0,0 +1,66 @@
+
+		result := strings.Replace(buf.String(), "\n", "<br>", -1)
+
+		resultTmpl.Execute(w, SearchResult{ Query:q, Result:template.HTML(result)})
```
Вёрстки здесь не должно быть, её можно сделать в шаблоне. Пример: https://github.com/polis-mail-ru-golang-1/examples/blob/master/lec6/blog/view/templates/posts.html

tariel-x commented on templates/result.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
базовые вещи можно вынести во вложенные шаблоны: https://github.com/polis-mail-ru-golang-1/examples/blob/master/lec6/blog/view/view.go#L21

tariel-x commented:
Добавьте ещё учёт знаков препинания, пожалуйста. Можно делать через strings.Trim или через strings.TrimFunc. Второе есть в examples. Первое - у кого-то в репозитории.

Тогда будет максимум баллов за 2-е дз.

r-i-besedin committed:
Логи, конфиг в новом пакете и улучшение шаблонов

Добавил логи к поиску, вынес конфиг в отдельный пакет, в шаблонах вынес базовые вещи во вложенные

r-i-besedin commented:
> Добавьте ещё учёт знаков препинания, пожалуйста. Можно делать через strings.Trim или через strings.TrimFunc. Второе есть в examples. Первое - у кого-то в репозитории.
> 
> Тогда будет максимум баллов за 2-е дз.

Хорошо, я сделаю это, но уже в 6 дз

r-i-besedin committed:
Update README.md

r-i-besedin committed:
Update README.md

tariel-x commented on invertedindex/invertedindex.go:
```diff
@@ -74,5 +75,23 @@ func SearchByWords(words []string) map[string]int {
 			resultmap[entry.Source]+=entry.Occurrence
 		}
 	}
+	log.Info().Msg("Поиск по строке "+strings.Join(words, " "))
```
Вот тут уместней было бы `log.Info().Msgf("%v", words)`

tariel-x commented on web/web.go:
```diff
@@ -0,0 +1,84 @@
+		"/res/",
+		http.FileServer(http.Dir("./static")),
+	))
+	mux.HandleFunc("/favicon.ico", func(w http.ResponseWriter, r *http.Request) {
```
можно в HTML указать, где находится иконка: `<link rel='icon' href='favicon.ico' type='image/x-icon'/ >` и тогда отдельный handler не понадобится.

tariel-x commented on invertedindex/invertedindex.go:
```diff
@@ -74,5 +75,23 @@ func SearchByWords(words []string) map[string]int {
+	sort.Sort(sort.Reverse(sort.IntSlice(a)))
+	for _, k := range a {
+		for _, s := range n[k] {
+			fmt.Fprintf(w, "- %s; совпадений - %d\n", s, k)
```
Почему бы всё таки не форматировать результат целиком в шаблоне? Получается, что всё равно у Вас часть логики отображения результатов определено в коде.

Можно завести структуру с результатами и в шаблоне выводить её так:
` - {{.File}} совпадений {{.Count}}<br>`
Это будет правильней в плане разделения логики: здесь формируются данные, там целиком формируется ответ.

Вдруг вы захотите переверстать страницу результатов, тогда с текущей реализацией ещё придётся искать по коду, где эти слова захардкожены. Или если понадобится сделать имя файла и количество жирным/курсивом - тогда придётся писать html теги в коде, а это так себе практика. Если вынести всё шаблон, такой ситуации не возникнет. Тем более сейчас результат формируется вообще в коде по-сути библиотеки, которую можно использовать отдельно в других проектах, возможно в таких, где никакой работы с пользователем не будет.

r-i-besedin committed:
Исправления по комментариям

Исправил строчку с логами, путь к иконке прописал в шаблоне, форматирование сделал в шаблоне. В пакете с индексом переименовал поле в структуре на weighted, чтобы можно было в будущем использовать более сложный расчет веса слова в файле, чем просто его повторение. Еще добавил структуры для веб и индекс файла, чтобы легче было перейти на MVC шаблон

r-i-besedin committed:
Забыл коммит

tariel-x commented:
:+1:
